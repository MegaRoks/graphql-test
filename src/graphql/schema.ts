import { buildSchema } from 'graphql';

export default buildSchema(`
    type BookType {
        bookId: Int!
        name: String!
        pageCount: Int!
        authorId: Int!
        author: AuthorType!
    }

    type AuthorType {
        authorId: Int!
        name: String!
    }

    type Query {
        getBooks: [BookType!]!
    }
`);
