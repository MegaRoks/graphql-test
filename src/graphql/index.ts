import { join } from 'path';
import { ApolloServer } from 'apollo-server';
import { buildSchema } from 'type-graphql';

export const getConnectionGraphQL = async () => {
    try {
        const schema = await buildSchema({ resolvers: [`${join(__dirname, '..', 'resolvers', '*.resolver.ts')}`] });
        return new ApolloServer({ schema });
    } catch (err) {
        console.error('err', err);
    }
};
