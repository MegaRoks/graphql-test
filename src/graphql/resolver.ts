import { Author } from './../entities/author.entity';
import connection from './../database';

const author = {
    authorId: 1,
    name: 'Author 1',
};

const books = [
    {
        bookId: 1,
        name: 'Book 1',
        pageCount: 22,
        authorId: 1,
        author,
    },
];

export default {
    async getBooks() {
        return connection
            .then((connect) => {
                const authorRepository = connect.getRepository(Author);
                return authorRepository.find();
            })
            .catch((err) => {
                console.error('error', err);
            });
    },
};
