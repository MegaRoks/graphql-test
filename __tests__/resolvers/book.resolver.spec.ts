import { makeExecutableSchema } from '@graphql-tools/schema';
import { addMocksToSchema } from '@graphql-tools/mock';
import { graphql } from 'graphql';

describe('Book Resolver Tests', () => {
    const schemaString = `
        type Author {
            authorId: String!
            name: String!
            createdAt: String!
            updatedAt: String!
        }

        type Book {
            bookId: String!
            name: String!
            pageCount: Int!
            authorId: String!
            createdAt: String!
            updatedAt: String!
            author: Author!
        }

        type Query {
            getBooksList: [Book!]!
            getBook(bookId: String!): Book!
        }

        input BookInput {
            name: String!
        }

        type Mutation {
            createBook(data: BookInput!): Book!
            updateBook(bookId: String!, data: BookInput!): Book!
            deleteBook(bookId: String!): Boolean!
        }
    `;

    const schema = makeExecutableSchema({ typeDefs: schemaString });
    const schemaWithMocks = addMocksToSchema({ schema });

    it('should get book list without author', async () => {
        const query = `
            query {
                getBooksList {
                    bookId name createdAt updatedAt
                }
            }
        `;

        const response = await graphql(schemaWithMocks, query);
        expect(response).toMatchSnapshot();
    });

    it('should get book list with author', async () => {
        const query = `
            query {
                getBooksList {
                    bookId name createdAt updatedAt author { authorId name createdAt updatedAt }
                }
            }
        `;

        const response = await graphql(schemaWithMocks, query);
        expect(response).toMatchSnapshot();
    });

    it('should get book', async () => {
        const query = `
            query {
                getBook(bookId: "qwe123") {
                    bookId name createdAt updatedAt
                }
            }
        `;

        const response = await graphql(schemaWithMocks, query);
        expect(response).toMatchSnapshot();
    });

    it('should create book', async () => {
        const query = `
            mutation {
                createBook(data: { name: "Book 1"}) {
                    bookId name createdAt updatedAt
                }
            }
        `;

        const response = await graphql(schemaWithMocks, query);
        expect(response).toMatchSnapshot();
    });

    it('should update book', async () => {
        const query = `
            mutation {
                updateBook(bookId: "qwe123", data: { name: "Book 2"}) {
                    bookId name createdAt updatedAt
                }
            }
        `;

        const response = await graphql(schemaWithMocks, query);
        expect(response).toMatchSnapshot();
    });

    it('should delete book', async () => {
        const query = `
            mutation {
                deleteBook(bookId: "qwe123")
            }
        `;

        const response = await graphql(schemaWithMocks, query);
        expect(response).toMatchSnapshot();
    });
});
