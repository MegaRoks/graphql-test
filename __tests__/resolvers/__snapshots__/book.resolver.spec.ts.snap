// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Resolver Tests should create book 1`] = `
Object {
  "data": Object {
    "createBook": Object {
      "bookId": "Hello World",
      "createdAt": "Hello World",
      "name": "Hello World",
      "updatedAt": "Hello World",
    },
  },
}
`;

exports[`Book Resolver Tests should delete book 1`] = `
Object {
  "data": Object {
    "deleteBook": true,
  },
}
`;

exports[`Book Resolver Tests should get book 1`] = `
Object {
  "data": Object {
    "getBook": Object {
      "bookId": "Hello World",
      "createdAt": "Hello World",
      "name": "Hello World",
      "updatedAt": "Hello World",
    },
  },
}
`;

exports[`Book Resolver Tests should get book list with author 1`] = `
Object {
  "data": Object {
    "getBooksList": Array [
      Object {
        "author": Object {
          "authorId": "Hello World",
          "createdAt": "Hello World",
          "name": "Hello World",
          "updatedAt": "Hello World",
        },
        "bookId": "Hello World",
        "createdAt": "Hello World",
        "name": "Hello World",
        "updatedAt": "Hello World",
      },
      Object {
        "author": Object {
          "authorId": "Hello World",
          "createdAt": "Hello World",
          "name": "Hello World",
          "updatedAt": "Hello World",
        },
        "bookId": "Hello World",
        "createdAt": "Hello World",
        "name": "Hello World",
        "updatedAt": "Hello World",
      },
    ],
  },
}
`;

exports[`Book Resolver Tests should get book list without author 1`] = `
Object {
  "data": Object {
    "getBooksList": Array [
      Object {
        "bookId": "Hello World",
        "createdAt": "Hello World",
        "name": "Hello World",
        "updatedAt": "Hello World",
      },
      Object {
        "bookId": "Hello World",
        "createdAt": "Hello World",
        "name": "Hello World",
        "updatedAt": "Hello World",
      },
    ],
  },
}
`;

exports[`Book Resolver Tests should update book 1`] = `
Object {
  "data": Object {
    "updateBook": Object {
      "bookId": "Hello World",
      "createdAt": "Hello World",
      "name": "Hello World",
      "updatedAt": "Hello World",
    },
  },
}
`;
